drivers/charger/charger_pf1550.c

```diff
@@ -0,0 +1,692 @@
+		return ret;
+	}
+
+	val = FIELD_GET(CHG_OPER_CHG_OPER_MASK, val);
+
+	switch (val) {
+	case PF1550_CHARGER_ON_LINEAR_ON:
+		*online = CHARGER_ONLINE_FIXED;
+		break;
+	default:
+		*online = CHARGER_ONLINE_OFFLINE;
+		break;
+	};
+
+	return 0;
+}
+
+static int pf1550_set_constant_charge_current(const struct device *dev, uint32_t current_ua)
+{
+	const struct charger_pf1550_config *const config = dev->config;
+	uint16_t idx;
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_group_get_index(charger_fast_charge_ua_range,
+					   ARRAY_SIZE(charger_fast_charge_ua_range), current_ua,
+					   &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = FIELD_PREP(CHG_CURR_CNFG_CHG_CC_MASK, idx);
+
+	return i2c_reg_update_byte_dt(&config->bus, CHARGER_CHG_CURR_CNFG,
+				      CHG_CURR_CNFG_CHG_CC_MASK, val);
+}
+
+static int pf1550_set_vbus_ilim(const struct device *dev, uint32_t current_ua)
+{
+	const struct charger_pf1550_config *const config = dev->config;
+	uint16_t idx;
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_group_get_index(charger_vbus_ilim_range,
+					   ARRAY_SIZE(charger_vbus_ilim_range), current_ua, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = FIELD_PREP(VBUS_INLIM_CNFG_VBUS_INLIM_MASK, idx);
+
+	return i2c_reg_update_byte_dt(&config->bus, CHARGER_VBUS_INLIM_CNFG,
+				      VBUS_INLIM_CNFG_VBUS_INLIM_MASK, val);
+}
+
+static int pf1550_set_vsys_min(const struct device *dev, uint32_t voltage_uv)
+{
+	const struct charger_pf1550_config *const config = dev->config;
+	uint16_t idx;
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_group_get_index(charger_vsysmin_uv, ARRAY_SIZE(charger_vsysmin_uv),
+					   voltage_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = FIELD_PREP(BATT_REG_VSYSMIN_MASK, idx);
+
+	return i2c_reg_update_byte_dt(&config->bus, CHARGER_BATT_REG, BATT_REG_VSYSMIN_MASK, val);
+}
+
+static int pf1550_set_charge_termination_uv(const struct device *dev, uint32_t voltage_uv)
+{
+	const struct charger_pf1550_config *const config = dev->config;
+	uint16_t idx;
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_group_get_index(charger_battery_termination_uv_range,
+					   ARRAY_SIZE(charger_battery_termination_uv_range),
+					   voltage_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = FIELD_PREP(BATT_REG_CHGCV_MASK, idx);
+
+	return i2c_reg_update_byte_dt(&config->bus, CHARGER_BATT_REG, BATT_REG_CHGCV_MASK, val);
+}
+
+static int pf1550_set_thermistor_mode(const struct device *dev, enum charger_pf1550_therm_mode mode)
+{
+	const struct charger_pf1550_config *const config = dev->config;
+	uint8_t val;
+
+	val = FIELD_PREP(THM_REG_CNFG_THM_CNFG_MASK, mode);
+
+	return i2c_reg_update_byte_dt(&config->bus, CHARGER_THM_REG_CNFG,
+				      THM_REG_CNFG_THM_CNFG_MASK, val);
+}
+
+static int pf1550_set_enabled(const struct device *dev, bool enable)
+{
+	struct charger_pf1550_data *data = dev->data;
+	const struct charger_pf1550_config *const config = dev->config;
+
+	data->charger_enabled = enable;
+
```

Comment:
Your cached state may be incorrect if the subsequent `i2c_reg_update_byte_dt` transaction fails.

Reply:
Fixed, thx